Modules/_asynciomodule.c

```diff
@@ -1963,7 +2135,10 @@ enter_task(asyncio_state *state, PyObject *loop, PyObject *task)
         Py_DECREF(item);
         return -1;
     }
-    Py_DECREF(item);
+
+    _PyThreadStateImpl *ts = (_PyThreadStateImpl *)_PyThreadState_GET();
+    assert(ts->asyncio_running_task == NULL);
+    ts->asyncio_running_task = item;  // strong ref
```

Comment:
This isn't safe. You can only do this when you can guarantee that running loop is the loop of this task.  This function can be called from any thread. 

Side question: IIUC all of this is only for debuggers. Why not just expose the current loop and let them read the current task from the current_tasks dict above, the dict offsets are already exposed.

Reply:
My idea which would work on all cases:
- Remove this tracking of current task altogether
- put the current_tasks dict on interpreter state
- expose the offset of that dict for external introspection
- read the current task with the help of current_tasks dict from interp offset and current loop from tstate offset.

Reply:
I working on implementing this.

Reply:
> Why not just expose the current loop and let them read the current task from the current_tasks dict above, the dict offsets are already exposed.

Because reading a dictionary is very complex externally. You cannot run Python code, you'd have to recreate the implementation on the other end, with all the pointer indirection.

Reply:
Kumar, this is a very good catch.

I agree that additional checks are warranted here. But I disagree that the current approach overall isn't valid. While technically "enter_task" *can* be called from some unrelated thread, in practice, it will only be called from one place -- around the Task's stepping function. For our asyncio Task that will always be the case, for third-party tasks that don't inherit from asyncio.Task I have to comments:

* we don't support tracking of third-party tasks with the new API anyway, so we just need to make sure we don't crash if a rogue task does highly questionable things;
* I can't imagine why would anyone on earth would call "asyncio.enter_task" from a different thread that their task is running in. Like it's just hard to even implement that, left alone have a sane motivation for that extra work. Or am I missing something? Please correct me if I'm wrong.

Therefore I'd argue that the current approach is OK, but I'm adding an additional check to the logic as well as an extensive comment elaborating on what's going on.

![CleanShot 2024-11-22 at 20 54 51@2x](https://github.com/user-attachments/assets/28065e2d-16b5-4036-a265-d56e30b29bde)

Reply:
> I can't imagine why would anyone on earth would call "asyncio.enter_task" from a different thread that their task is running in. Like it's just hard to even implement that, left alone have a sane motivation for that extra work. Or am I missing something? Please correct me if I'm wrong.

This may happen if in case someone is running two separate event loops in two threads and tries to introspect the call stack of the other. I understand that this may seem far fetch but still better safe than sorry. Pushed a minor suggestion to your comment.

Reply:
> This may happen if in case someone is running two separate event loops in two threads and tries to introspect the call stack of the other. I understand that this may seem far fetch but still better safe than sorry. Pushed a minor suggestion to your comment.

That will work, introspection doesn't rely on `enter_task` itself. So, again, I don't see how what you're saying is possible at all.